sewer/auth.py

```diff
@@ -1,42 +1,115 @@
+    delete_dns_record are obsolescent, and will be removed when there are no
+    provders left in the tree that use them.
+
+    The implementation of the *_cert methods here just loops, calling *_auth for
+    each item in the authprizations list.  So providers that have to deal with one
+    auth at a time can just implement that.  Some providers may be able to work
+    more efficently if they can see all the items at once, and those will prefer
+    to implement *_cert, overriding the versions here.
+
+    __init__ arguments protocol:  Taking the lesson from some Django classes that
+    were layered like this (and with many leaf users, some outside of the framework),
+    it is always declared with only self and **kwargs.  Each layer MUST pop the args
+    it knows and wishes to handle, then pass what's left to its immediate base class.
+    The end goal is to allow initializing a provider with whatever it needs to
+    authenticate to and interact with the DNS system that publishes the challenges.
+    Then that can be done outside of sewer's ACME engine (Client class) and passed
+    to it ready to use.
+    """
+
+    def __init__(self, **kwargs):
         """
-        called after the cert is aquired to cleanup any authorization resources created in the fulfill_authorization method above.
+        chal_types      Sequence[str] set of implemented ACME challenge types
+
+        optional logging control argument(s).  From preferred to legacy:
+         *  logger	a configured logging object to use as-is
+         *  log_level	level to setup local logger for, defaulting to INFO
+         *  LOG_LEVEL   obsolescent legacy version of log_level
+
+        logger will override log_level will override LOG_LEVEL if more than one is passed
         """
-        raise NotImplementedError("cleanup_authorization method must be implemented.")
+
+        ### FIX ME ### auth_type handling?  and that's not best named, should be challenge_type
+
+        self.chal_types = kwargs.pop("chal_types")
+        if not all(ct in supported_validations for ct in self.chal_types):
+            raise ValueError("Unsupported challenge type in list: %s" % self.chal_types)
+
+        logger = kwargs.pop("logger", None)
+        log_level = kwargs.pop("LOG_LEVEL", "INFO")  # compatibility / obsolescent
+        log_level = kwargs.pop("log_level", log_level)
+        if kwargs:
+            raise ValueError("BaseAuth was passed unrecognized argument(s): %s" % kwargs)
+
+        if logger:
+            self.logger = logger
+        else:
+            self.logger = logging.getLogger(__name__)
+            self.logger.setLevel(log_level)
+            if not self.logger.hasHandlers():
+                handler = logging.StreamHandler()
+                formatter = logging.Formatter("%(message)s")
+                handler.setFormatter(formatter)
+                self.logger.addHandler(handler)
+
+    def setup_cert(self, authorizations: Sequence[Tuple[str, str, str]]):
```

Comment:
I'm not sure `setup_cert` and `clear_cert` are the right names for these methods, but I have nothing better right now. I'll need to give it more thought. 

also will this method need to return the `cleanup_list` used in client.py? something like this?

```python
def setup_cert(self, authorizations: Sequence[Tuple[str, str, str]]) ->  Sequence[Tuple[str, str, str]]:
    return [ self.setup_auth(*auth) for auth in authorizations ]
```

Reply:
after a bit more thought, why not only have clients implement the multiple authorizations methods. something like this..

```python
# previously called setup_cert
def setup(self, authorizations: Sequence[Tuple[str, str, str]]) -> Sequence[Tuple[str, str, str]]:
     "setup list of authorizations - add dns records, files, etc."
     raise NotImplementedError("setup method must be implemented")

# previously called cleanup_cert
def cleanup(self, authorizations: Sequence[Tuple[str, str, str]]):
     "cleanup list of authorizations - removes dns records, files, etc."
     raise NotImplementedError("cleanup method must be implemented")
````

the clients can break that down task however they want (batched together, per auth, etc). One way to look at it would be that this base class just defines a common API for `client.py` to use

Reply:
3-in-1 reply!

Using setup_cert is on the todo list - that just requires collecting all the challenges instead of firing them off one by one.  And no, they no longer have to return a filtered list - that was delegating work to the driver that the client code can do in one line!

And yes, the endpoint is that drivers all accept the *_cert calls.  Everything else is for compatibility with existing code (everything in dns_providers).

Oh.  I adopted the setup and clear prefixes (did I change one of those along the way?), and when I was looking for the name suffix I realized that the batches are going to naturally be the challenges for one certificate request, and that sounded better to me than whatever alternatives I tried.

Reply:
So while looking at the next step in reworking this series as if I'd known what I was doing <wink>, I found out where my choice of prefixes came from.  You gave me _cleanup_, and perhaps because I'd recently been mucking about in some existing tests, _setup_ seemed the best complement to it.

I knew I'd stolen them from somewhere!